### PR TITLE
Allow setting VSI credentials when loading OGR/GDAL layers

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR exposes the VSI credential options for user control when adding OGR vector/GDAL raster layers from the Data Source Manager, eg:

image

The underlying provider changes were discussed in https://github.com/qgis/QGIS/pull/57801. When credential options are found in a layer's URI, we use GDAL's VSISetPathSpecificOption to set the credential option for that VSI driver and bucket. 

This allows per-vsi driver & bucket credentials for layers, whereas other approaches (like environment variable setting) force a single set of credentials to be used for an entire QGIS session. (And are difficult for users to set)

I've also taken the opportunity to cleanup a bunch of code relating to how VSI protocols are exposed in GUI, and how GUIs for GDAL configuration options are dynamically created.

Requires GDAL 3.5+

### CODE REVIEW

**Comment by nirvn** [comment]
@nyalldawson , that's really nice. The limited one credential (via environment variables) was making it really hard to enjoy cloud hosted datasets before

File: src/gui/ogr/qgsogrhelperfunctions.h [hunk 1]
```diff
@@ -19,18 +19,42 @@
...
- * \note not available in python bindings
+ * \ingroup gui
+ * \class QgsGdalGuiUtils
+ * \brief Utility functions for working with GDAL in GUI classes.
+ *
+ * \note not available in Python bindings
+ * \since QGIS 3.40
  */
-QString GUI_EXPORT createDatabaseURI( const QString &connectionType, const QString &host, const QString &database, QString port, const QString &configId, QString username, QString password, bool expandAuthConfig = false );
+class GUI_EXPORT QgsGdalGuiUtils
```
Comment (rouault) [inline src/gui/ogr/qgsogrhelperfunctions.h:35]: shouldn't the file be renamed to qgsgdalguiutils.h for consistency?

**Review by rouault — APPROVED** [review]

File: src/gui/ogr/qgsogrhelperfunctions.h (hunk 1 above)
Comment (nyalldawson) [reply src/gui/ogr/qgsogrhelperfunctions.h:35]: Yes, it should! I'll do that in a follow up PR I've got in progress to avoid merge conflicts.

Comment (nyalldawson) [reply src/gui/ogr/qgsogrhelperfunctions.h:35]: Done in https://github.com/qgis/QGIS/pull/57868

**Pull request merged**